src/wallet/hdwallet.ts

```diff
@@ -44,57 +41,50 @@ export abstract class HDWallet {
+		return `m/44'/${coinType}'`;
+	}
+	
+	protected coinType(): number {
+		return (this.constructor as typeof AbstractHDWallet)._coinType!;
+	}
 
     // helper methods that adds an account HD node to the HD wallet following the BIP-44 standard.
     protected addAccount(accountIndex: number): void {
         const newNode = this._root.deriveChild(accountIndex);
         this._accounts.set(accountIndex, newNode);
     }
 
-    protected deriveAddress(
-        account: number,
-        startingIndex: number,
-        zone: string,
-        isChange: boolean = false,
-    ): HDNodeWallet {
+    protected deriveAddress(account: number, startingIndex: number, zone: Zone, isChange: boolean = false): HDNodeWallet {
+        this.validateZone(zone);
         const isValidAddressForZone = (address: string) => {
-            const zoneByte = getZoneForAddress(address);
-            if (!zone) {
+            const addressZone = getZoneForAddress(address);
+            if (!addressZone) {
                 return false;
             }
-            const shardNickname = ZoneData.find((zoneData) => zoneData.byte === zoneByte)?.nickname;
-            const isCorrectShard = shardNickname === zone.toLowerCase();
-            const isCorrectLedger = this.coinType() === 969 ? isQiAddress(address) : !isQiAddress(address);
-            return isCorrectShard && isCorrectLedger;
-        };
-        // derive the address node
-        const accountNode = this._accounts.get(account);
-        const changeIndex = isChange ? 1 : 0;
-        const changeNode = accountNode!.deriveChild(changeIndex);
-        let addrIndex: number = startingIndex;
-        let addressNode: HDNodeWallet;
-        do {
-            addressNode = changeNode.deriveChild(addrIndex);
-            addrIndex++;
-            // put a hard limit on the number of addresses to derive
-            if (addrIndex - startingIndex > MAX_ADDRESS_DERIVATION_ATTEMPTS) {
-                throw new Error(
-                    `Failed to derive a valid address for the zone ${zone} after MAX_ADDRESS_DERIVATION_ATTEMPTS attempts.`,
-                );
-            }
-        } while (!isValidAddressForZone(addressNode.address));
+			const isCorrectShard = addressZone === zone;
+			const isCorrectLedger = (this.coinType() === 969) ? isQiAddress(address) : !isQiAddress(address);
+			return isCorrectShard && isCorrectLedger;
+		}
+		// derive the address node
+		const accountNode = this._accounts.get(account);
+		const changeIndex = isChange ? 1 : 0;
+		const changeNode = accountNode!.deriveChild(changeIndex);
+		let addrIndex: number = startingIndex;
+		let addressNode: HDNodeWallet;
+		do {
+			addressNode = changeNode.deriveChild(addrIndex);
+			addrIndex++;
+			// put a hard limit on the number of addresses to derive
+			if (addrIndex - startingIndex > MAX_ADDRESS_DERIVATION_ATTEMPTS) {
+				throw new Error(`Failed to derive a valid address for the zone ${zone} after MAX_ADDRESS_DERIVATION_ATTEMPTS attempts.`);
```

Comment:
Let's log the actual value of `MAX_ADDRESS_DERIVATION_ATTEMPTS` by wrapping in `${}` block